searx/network/context.py

```diff
@@ -0,0 +1,404 @@
+
+Various implementations define what to do when there is an exception in the function `my_engine`:
+
+* `NetworkContextRetryFunction` gets another HTTP client and tries the whole function again.
+* `NetworkContextRetryDifferentHTTPClient` gets another HTTP client and tries the query again.
+* `NetworkContextRetrySameHTTPClient` tries the query again with the same HTTP client.
+"""
+import functools
+import ssl
+from abc import ABC, abstractmethod
+from contextlib import contextmanager
+from timeit import default_timer
+from typing import Callable, Optional, final
+
+try:
+    from typing import ParamSpec, TypeVar
+except ImportError:
+    # to support Python < 3.10
+    from typing_extensions import ParamSpec, TypeVar
+
+import httpx
+
+from searx.network.client import ABCHTTPClient, SoftRetryHTTPException
+
+P = ParamSpec('P')
+R = TypeVar('R')
+HTTPCLIENTFACTORY = Callable[[], ABCHTTPClient]
+
+DEFAULT_TIMEOUT = 120.0
+
+
+## NetworkContext
+
+
+class NetworkContext(ABC):
+    """Abstract implementation: the call must defined in concrete classes.
+
+    Lifetime: one engine request or initialization of an engine.
+    """
+
+    __slots__ = ('_retries', '_http_client_factory', '_http_client', '_start_time', '_http_time', '_timeout')
+
+    def __init__(
+        self,
+        retries: int,
+        http_client_factory: HTTPCLIENTFACTORY,
+        start_time: Optional[float],
+        timeout: Optional[float],
+    ):
+        self._retries: int = retries
+        # wrap http_client_factory here, so we can forget about this wrapping
+        self._http_client_factory = _TimeHTTPClientWrapper.wrap_factory(http_client_factory, self)
+        self._http_client: Optional[ABCHTTPClient] = None
+        self._start_time: float = start_time or default_timer()
+        self._http_time: float = 0.0
+        self._timeout: Optional[float] = timeout
+
+    @abstractmethod
+    def call(self, func: Callable[P, R], *args: P.args, **kwargs: P.kwargs) -> R:
+        """Call func within the network context.
+
+        The retry policy might call func multiple times.
+
+        Within the function self.get_http_client() returns an HTTP client to use.
+
+        The retry policy might send multiple times the same HTTP request
+        until it works or the retry count falls to zero.
+        """
+
+    @final
+    def request(self, *args, **kwargs):
+        """Convenient method to wrap a call to request inside the call method.
+
+        Use a new HTTP client to wrap a call to the request method using self.call
+        """
+        self._set_http_client()
+        return self.call(self._get_http_client().request, *args, **kwargs)
+
+    @final
+    def stream(self, *args, **kwargs):
+        """Convenient method to wrap a call to stream inside the call method.
+
+        Use a new HTTP client to wrap a call to the stream method using self.call
+        """
+        self._set_http_client()
```

Comment:
@dalf  Why is a new client set here? / Actually, the retry strategy (the `self.call`) is already responsible for setting a new client. By example in line 250 a `self._set_http_client()` is set by the `call` method.

```python

class NetworkContextRetryFunction(NetworkContext):
    ...
    def call(self, func: Callable[P, R], *args: P.args, **kwargs: P.kwargs) -> R:
        try:
            while self._retries >= 0 and self.get_remaining_time() > 0:
                self._set_http_client()

```

Reply:
I know: an explanation or a different implementation is needed here.
You will see it crashes if you remove this line, I don't remember why right now. I will look at it.

Reply:
> I will look at it later this week / this day.

No need to hurry, please let me first finish my review .. in this phase I will post some of my question here on gh .. if needed, we can comeback later to this topics/quesions.

Reply:
Fixed in ed35731 
The call to `self._get_http_client()` has to be done inside `self.call`, not before.